cpp/include/cudf/algorithm/scan_copy_if.cuh

```diff
@@ -0,0 +1,442 @@
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_scalar.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/iterator/transform_output_iterator.h>
+
+#include <cub/agent/single_pass_scan_operators.cuh>
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_scan.cuh>
+
+#include <iterator>
+#include <type_traits>
+
+template <typename Dividend, typename Divisor>
+inline constexpr auto ceil_div(Dividend dividend, Divisor divisor)
+{
+  return dividend / divisor + (dividend % divisor != 0);
+}
+
+// ===== Agent =====================================================================================
+
+template <typename Policy>
+struct agent {
+  typename Policy::InputIterator d_input;
+  typename Policy::OutputCountIterator d_output_count;
+  typename Policy::OutputIterator d_output;
+  uint32_t num_items;
+  typename Policy::ScanOperator scan_op;
+  typename Policy::PredOperator pred_op;
+  typename Policy::ItemsTileState& items_tile_state;
+  typename Policy::IndexTileState& index_tile_state;
+
+  inline __device__ void consume_range(bool const do_scatter,
+                                       uint32_t const num_tiles,
+                                       uint32_t const start_tile)
+  {
+    uint32_t const tile_idx            = start_tile + blockIdx.x;
+    uint32_t const tile_offset         = Policy::ITEMS_PER_TILE * tile_idx;
+    uint32_t const num_items_remaining = num_items - tile_offset;
+
+    typename Policy::OutputCount num_selected;
+
+    if (tile_idx < num_tiles - 1) {
+      num_selected = consume_tile<false>(do_scatter, tile_idx, tile_offset, Policy::ITEMS_PER_TILE);
+    } else {
+      num_selected = consume_tile<true>(do_scatter, tile_idx, tile_offset, num_items_remaining);
+      if (threadIdx.x == 0) { *d_output_count = num_selected; }
+    }
+  }
+
+  template <bool IS_LAST_TILE>
+  inline __device__ uint32_t consume_tile(bool const do_scatter,
+                                          uint32_t const tile_idx,
+                                          uint32_t const tile_offset,
+                                          uint32_t const num_items_remaining)
+  {
+    typename Policy::Input items[Policy::ITEMS_PER_THREAD];
+
+    __shared__ union {
+      typename Policy::ItemsBlockLoad::TempStorage item_load;
+      struct {
+        typename Policy::IndexBlockScan::TempStorage index_scan;
+        typename Policy::IndexPrefixCallback::TempStorage index_prefix;
+      };
+      struct {
+        typename Policy::ItemsBlockScan::TempStorage items_scan;
+        typename Policy::ItemsPrefixCallback::TempStorage items_prefix;
+      };
+    } temp_storage;
+
+    // Load Inputs
+
+    if (IS_LAST_TILE) {
+      Policy::ItemsBlockLoad(temp_storage.item_load)  //
+        .Load(d_input + tile_offset, items, num_items_remaining);
+    } else {
+      Policy::ItemsBlockLoad(temp_storage.item_load)  //
+        .Load(d_input + tile_offset, items);
+    }
+
+    __syncthreads();
```

Comment:
Hadn't heard of cooperatives groups until now. Seems useful for keeping things flexible or clearly expressing intent, but I'm having a hard time seeing how they'd improve this code. The CUB functions being called all `__sync...` or `__threadfence...` internally, so adding a `thread_block_tile<32>` parameter to this function seems like it would actually be more misleading due to inconsistent paradigms. I should make this function private...